File: core/src/main/java/com/shatteredpixel/shatteredpixeldungeon/levels/rooms/standard/StandardRoom.java
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2015 Oleg Dolya
 *
 * Shattered Pixel Dungeon
 * Copyright (C) 2014-2022 Evan Debenham
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.shatteredpixel.shatteredpixeldungeon.levels.rooms.standard;

import com.shatteredpixel.shatteredpixeldungeon.Dungeon;
import com.shatteredpixel.shatteredpixeldungeon.levels.Level;
import com.shatteredpixel.shatteredpixeldungeon.levels.Terrain;
import com.shatteredpixel.shatteredpixeldungeon.levels.rooms.Room;
import com.watabou.utils.Point;
import com.watabou.utils.Random;
import com.watabou.utils.Reflection;

import java.util.ArrayList;

public abstract class StandardRoom extends Room {
	
	public enum SizeCategory {
		
		NORMAL(4, 10, 1),
		LARGE(10, 14, 2),
		GIANT(14, 18, 3);
		
		public final int minDim, maxDim;
		public final int roomValue;
		
		SizeCategory(int min, int max, int val){
			minDim = min;
			maxDim = max;
			roomValue = val;
		}
		
		public int connectionWeight(){
			return roomValue*roomValue;
		}
		
	}
	
	public SizeCategory sizeCat;
	{ setSizeCat(); }
	
	//Note that if a room wishes to allow itself to be forced to a certain size category,
	//but would (effectively) never roll that size category, consider using Float.MIN_VALUE
	public float[] sizeCatProbs(){
		//always normal by default
		return new float[]{1, 0, 0};
	}
	
	public boolean setSizeCat(){
		return setSizeCat(0, SizeCategory.values().length-1);
	}
	
	//assumes room value is always ordinal+1
	public boolean setSizeCat( int maxRoomValue ){
		return setSizeCat(0, maxRoomValue-1);
	}
	
	//returns false if size cannot be set
	public boolean setSizeCat( int minOrdinal, int maxOrdinal ) {
		float[] probs = sizeCatProbs();
		SizeCategory[] categories = SizeCategory.values();
		
		if (probs.length != categories.length) return false;
		
		for (int i = 0; i < minOrdinal; i++)                    probs[i] = 0;
		for (int i = maxOrdinal+1; i < categories.length; i++)  probs[i] = 0;
		
		int ordinal = Random.chances(probs);
		
		if (ordinal != -1){
			sizeCat = categories[ordinal];
			return true;
		} else {
			return false;
		}
	}
	
	@Override
	public int minWidth() { return sizeCat.minDim; }
	public int maxWidth() { return sizeCat.maxDim; }
	
	@Override
	public int minHeight() { return sizeCat.minDim; }
	public int maxHeight() { return sizeCat.maxDim; }

	@Override
	public boolean canMerge(Level l, Point p, int mergeTerrain) {
		int cell = l.pointToCell(pointInside(p, 1));
		return (Terrain.flags[l.map[cell]] & Terrain.SOLID) == 0;
	}

	//FIXME this is a very messy way of handing variable standard rooms
	private static ArrayList<Class<?extends StandardRoom>> rooms = new ArrayList<>();
	static {
		rooms.add(EmptyRoom.class);


		rooms.add(SewerPipeRoom.class);
		rooms.add(RingRoom.class);
		rooms.add(CircleBasinRoom.class);

		rooms.add(SegmentedRoom.class);
		rooms.add(PillarsRoom.class);
		rooms.add(CellBlockRoom.class);

		rooms.add(CaveRoom.class);
		rooms.add(CavesFissureRoom.class);
		rooms.add(CirclePitRoom.class);

		rooms.add(HallwayRoom.class);
		rooms.add(StatuesRoom.class);
		rooms.add(SegmentedLibraryRoom.class);

		rooms.add(RuinsRoom.class);
		rooms.add(ChasmRoom.class);
		rooms.add(SkullsRoom.class);


		rooms.add(PlantsRoom.class);
		rooms.add(AquariumRoom.class);
		rooms.add(PlatformRoom.class);
		rooms.add(BurnedRoom.class);
		rooms.add(FissureRoom.class);
		rooms.add(GrassyGraveRoom.class);
		rooms.add(StripedRoom.class);
		rooms.add(StudyRoom.class);
		rooms.add(SuspiciousChestRoom.class);
		rooms.add(MinefieldRoom.class);
	}
	
	private static float[][] chances = new float[27][];
	static {
		chances[1] =  new float[]{10,  10,10,5, 0,0,0, 0,0,0, 0,0,0, 0,0,0,  1,0,1,0,1,0,1,1,0,0};
		chances[2] =  new float[]{10,  10,10,5, 0,0,0, 0,0,0, 0,0,0, 0,0,0,  1,1,1,1,1,1,1,1,1,1};
		chances[4] =  chances[3] = chances[2];
		chances[5] =  new float[]{10,  10,10,0, 0,0,0, 0,0,0, 0,0,0, 0,0,0,  0,0,0,0,0,0,0,0,0,0};

		chances[6] =  new float[]{10,  0,0,0, 10,10,5, 0,0,0, 0,0,0, 0,0,0,  1,1,1,1,1,1,1,1,1,1};
		chances[10] = chances[9] = chances[8] = chances[7] = chances[6];

		chances[11] = new float[]{10,  0,0,0, 0,0,0, 10,10,5, 0,0,0, 0,0,0,  1,1,1,1,1,1,1,1,1,1};
		chances[15] = chances[14] = chances[13] = chances[12] = chances[11];

		chances[16] = new float[]{10,  0,0,0, 0,0,0, 0,0,0, 10,10,5, 0,0,0,  1,1,1,1,1,1,1,1,1,1};
		chances[20] = chances[19] = chances[18] = chances[17] = chances[16];

		chances[21] = new float[]{10,  0,0,0, 0,0,0, 0,0,0, 0,0,0, 10,10,5,  1,1,1,1,1,1,1,1,1,1};
		chances[26] = chances[25] = chances[24] = chances[23] = chances[22] = chances[21];
	}
	
	
	public static StandardRoom createRoom(){
		return Reflection.newInstance(rooms.get(Random.chances(chances[Dungeon.depth])));
	}
	
}

File: pixel-dungeon-web/src/levels/rooms/StandardRoom.js
/*
 * Pixel Dungeon Web Port
 * Ported from Shattered Pixel Dungeon
 * Copyright (C) 2012-2015 Oleg Dolya
 * Copyright (C) 2014-2025 Evan Debenham
 *
 * Standard Procedural Room
 * Source: core/src/main/java/com/shatteredpixel/shatteredpixeldungeon/levels/rooms/standard/StandardRoom.java
 */

import { Room } from './Room.js';
import { Random } from '../../utils/Random.js';
import { TerrainType, flags as TerrainFlags } from '../Terrain.js';
import { Painter } from '../painters/Painter.js';

/**
 * Room size categories
 * Source: StandardRoom.java:36-51
 */
export const SizeCategory = {
    NORMAL: { minDim: 4, maxDim: 10, roomValue: 1 },
    LARGE:  { minDim: 10, maxDim: 14, roomValue: 2 },
    GIANT:  { minDim: 14, maxDim: 18, roomValue: 3 }
};

/**
 * Base class for standard procedural rooms
 * Source: StandardRoom.java:34-194
 */
export class StandardRoom extends Room {
    constructor() {
        super();
        this.sizeCat = null;
        this.setSizeCat();
    }

    /**
     * Probability weights for each size category
     * Source: StandardRoom.java:58-61
     */
    sizeCatProbs() {
        // Always normal by default
        return [1, 0, 0];
    }

    /**
     * Set size category
     * Source: StandardRoom.java:63-90
     */
    setSizeCat(minOrdinal = 0, maxOrdinal = Object.keys(SizeCategory).length - 1) {
        const probs = this.sizeCatProbs();
        const categories = Object.keys(SizeCategory);

        if (probs.length !== categories.length) return false;

        // Zero out probabilities outside the valid range
        for (let i = 0; i < minOrdinal; i++) probs[i] = 0;
        for (let i = maxOrdinal + 1; i < categories.length; i++) probs[i] = 0;

        const ordinal = Random.chances(probs);

        if (ordinal !== -1) {
            this.sizeCat = SizeCategory[categories[ordinal]];
            return true;
        } else {
            return false;
        }
    }

    /**
     * Set size category by max room value
     * Source: StandardRoom.java:68-70
     */
    setSizeCat_maxValue(maxRoomValue) {
        return this.setSizeCat(0, maxRoomValue - 1);
    }

    // ========== Size constraints ==========
    // Source: StandardRoom.java:92-98

    minWidth() { return this.sizeCat.minDim; }
    maxWidth() { return this.sizeCat.maxDim; }
    minHeight() { return this.sizeCat.minDim; }
    maxHeight() { return this.sizeCat.maxDim; }

    /**
     * Size factor for counting purposes
     * Larger rooms count as multiple rooms
     * Source: StandardRoom.java:102-104
     */
    sizeFactor() {
        return this.sizeCat.roomValue;
    }

    /**
     * Mob spawn weight
     * Source: StandardRoom.java:106-111
     */
    mobSpawnWeight() {
        if (this.isEntrance()) {
            return 1; // Entrance rooms don't have higher mob spawns
        }
        return this.sizeFactor();
    }

    /**
     * Connection weight for graph building
     * Source: StandardRoom.java:113-115
     */
    connectionWeight() {
        return this.sizeFactor() * this.sizeFactor();
    }

    /**
     * Can merge with another room?
     * Source: StandardRoom.java:118-121
     */
    canMerge(level, other, p, mergeTerrain) {
        const cell = level.pointToCell(this.pointInside(p, 1));
        return (TerrainFlags.flags[level.map[cell]] & TerrainFlags.SOLID) === 0;
    }

    /**
     * Basic paint implementation
     * Source: EntranceRoom.java:77-100 (used as reference)
     */
    paint(level) {
        // Default empty implementation - subclasses override
        throw new Error('StandardRoom.paint() must be overridden by subclass');
    }

    /**
     * Create a random standard room
     * Source: StandardRoom.java:190-192
     *
     * SIMPLIFIED: For initial port, returns a basic EmptyRoom
     * Full implementation would select from a list of specific room types
     */
    static createRoom() {
        return new EmptyRoom();
    }
}

/**
 * Simple empty room (for initial testing)
 * A basic rectangular room with floor tiles
 */
class EmptyRoom extends StandardRoom {
    constructor() {
        super();
    }

    /**
     * Paint a simple rectangular room
     */
    paint(level) {
        // Fill with floor tiles
        Painter.fill(level, this, TerrainType.EMPTY);
    }
}
